db/dcrpg/pgblockchain.go

```diff
@@ -2833,93 +2782,50 @@ func (pgb *ChainDB) getChartData(data map[string]*dbtypes.ChartsData,
-	if err != nil {
-		return
-	}
-
-	chainData := [2]*dbtypes.ChartsData{
-		data[dbtypes.ChainWork],
-		data[dbtypes.HashRate],
-	}
-
-	chainData, err = pgb.chainWork(chainData)
-	if err != nil {
-		return
-	}
+func (pgb *ChainDB) PgChartsData(charts *cache.ChartData) (err error) {
+	lastHeight, t := charts.TipStats()
+	since := time.Unix(int64(t), 0)
 
-	avgSize := pgb.getChartData(data, dbtypes.AvgBlockSize)
-	blockSize := pgb.getChartData(data, dbtypes.BlockChainSize)
-	blockSize.Time, blockSize.ChainSize, avgSize.Size, err = pgb.blocksByTime(blockSize.Time,
-		blockSize.ChainSize, avgSize.Size)
+	err = pgb.blockStats(lastHeight, charts)
 	if err != nil {
 		return
 	}
 
-	avgSize.Time = blockSize.Time
-
-	durPerB := pgb.getChartData(data, dbtypes.DurationBTW)
-	txPerB := pgb.getChartData(data, dbtypes.TxPerBlock)
-	txPerB.Height, txPerB.Count, durPerB.ValueF, err = pgb.blocksByHeight(txPerB.Height,
-		txPerB.Count, durPerB.ValueF)
-	if err != nil {
+	_, t = charts.TipStats()
+	if since.Equal(time.Unix(int64(t), 0)) {
+		// nothing was updated, log a warning, but no error
+		log.Warnf("PgChartsData: no new data found")
 		return
 	}
 
-	durPerB.Height = txPerB.Height
-
-	tickets := pgb.getChartData(data, dbtypes.TicketPrice)
-	diff := pgb.getChartData(data, dbtypes.POWDifficulty)
-	tickets.Time, tickets.ValueF, diff.Difficulty, err = pgb.ticketsPriceByHeight(tickets.Time,
-		tickets.ValueF, diff.Difficulty)
+	err = pgb.coinSupply(lastHeight, charts)
 	if err != nil {
 		return
 	}
 
-	diff.Time = tickets.Time
+	err = pgb.windowStats(pgb.chainParams.StakeDiffWindowSize, lastHeight,
+		charts)
 
-	tByAllBlocks := pgb.getChartData(data, dbtypes.TicketsByBlocks)
-	tByAllBlocks.Height, tByAllBlocks.Solo, tByAllBlocks.Pooled, err = pgb.ticketsByBlocks(tByAllBlocks.Height,
-		tByAllBlocks.Solo, tByAllBlocks.Pooled)
-	if err != nil {
-		return
-	}
-
-	ticketsByWin := pgb.getChartData(data, dbtypes.TicketByWindows)
-	ticketsByWin.Height, ticketsByWin.Solo, ticketsByWin.Pooled, err = pgb.ticketsByTPWindows(ticketsByWin.Height,
-		ticketsByWin.Solo, ticketsByWin.Pooled)
-	if err != nil {
-		return
-	}
-
-	ticketSpendT := pgb.getChartData(data, dbtypes.TicketSpendT)
-	ticketSpendT.Height, ticketSpendT.Unspent, ticketSpendT.Revoked, err = pgb.ticketSpendTypePerBlock(ticketSpendT.Height,
-		ticketSpendT.Unspent, ticketSpendT.Revoked)
-	if err != nil {
-		return
-	}
-
-	data[dbtypes.AvgBlockSize] = avgSize
-	data[dbtypes.BlockChainSize] = blockSize
-	data[dbtypes.ChainWork] = chainData[0]
-	data[dbtypes.CoinSupply] = supply
-	data[dbtypes.DurationBTW] = durPerB
-	data[dbtypes.HashRate] = chainData[1]
-	data[dbtypes.POWDifficulty] = diff
-	data[dbtypes.TicketByWindows] = ticketsByWin
-	data[dbtypes.TicketPrice] = tickets
-	data[dbtypes.TicketsByBlocks] = tByAllBlocks
-	data[dbtypes.TicketSpendT] = ticketSpendT
-	data[dbtypes.TxPerBlock] = txPerB
-	data[dbtypes.TxPerDay] = txRate
+	// tByAllBlocks := pgb.getChartData(data, dbtypes.TicketsByBlocks)
```

Comment:
There are a couple of data sets that I haven't re-implemented. The solo/pooled tickets and unspent/revoked tickets by purchase block. I question their value, but I am happy to grab that data if requested. The solo/pooled accounting in particular has come up as innacurate due to ticket splitting.